services/sharedchannel/sync_recv.go

```diff
@@ -28,223 +28,211 @@ func (scs *Service) onReceiveSyncMessage(msg model.RemoteClusterMsg, rc *model.R
-				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error getting Team for Channel",
-					mlog.String("ChannelId", sm.Post.ChannelId),
-					mlog.String("PostId", sm.Post.Id),
-					mlog.Err(err2),
-				)
-				postErrors = append(postErrors, sm.Post.Id)
-				continue
-			}
-			chanToTeamMap[sm.Post.ChannelId] = team
-		}
 
-		sm.Post.Message = scs.processPermalinkFromRemote(sm.Post, chanToTeamMap[sm.Post.ChannelId])
+		scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "Sync msg received",
+			mlog.String("post_id", sm.PostId),
+			mlog.String("channel_id", sm.ChannelId),
+			mlog.Int("reaction_count", len(sm.Reactions)),
+			mlog.Int("user_count", len(sm.Users)),
+			mlog.Bool("has_post", sm.Post != nil),
+		)
 
 		if channel == nil {
-			if channel, err = scs.server.GetStore().Channel().Get(sm.Post.ChannelId, true); err != nil {
+			if channel, err = scs.server.GetStore().Channel().Get(sm.ChannelId, true); err != nil {
 				// if the channel doesn't exist then none of these sync messages are going to work.
-				return postErrors, 0, fmt.Errorf("channel not found processing sync messages: %w", err)
+				return fmt.Errorf("channel not found processing sync messages: %w", err)
+			}
+		}
+
+		if sm.Post != nil {
+			if _, ok := chanToTeamMap[sm.Post.ChannelId]; !ok {
+				team, err2 := scs.server.GetStore().Channel().GetTeamForChannel(sm.Post.ChannelId)
+				if err2 != nil {
+					scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error getting Team for Channel",
+						mlog.String("ChannelId", sm.Post.ChannelId),
+						mlog.String("PostId", sm.Post.Id),
+						mlog.Err(err2),
+					)
+					postErrors = append(postErrors, sm.Post.Id)
+					continue
+				}
+				chanToTeamMap[sm.Post.ChannelId] = team
+			}
+			sm.Post.Message = scs.processPermalinkFromRemote(sm.Post, chanToTeamMap[sm.Post.ChannelId])
+		}
+
+		// add/update users first
+		for _, user := range sm.Users {
+			if userSaved, err := scs.upsertSyncUser(user, rc); err != nil {
+				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error upserting sync user",
+					mlog.String("post_id", sm.PostId),
+					mlog.String("channel_id", sm.ChannelId),
+					mlog.String("user_id", user.Id),
+					mlog.Err(err))
+			} else {
+				usersSyncd = append(usersSyncd, userSaved.Id)
+				scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceDebug, "User upserted via sync",
+					mlog.String("channel_id", sm.ChannelId),
+					mlog.String("user_id", user.Id))
 			}
 		}
 
-		if err := scs.server.GetStore().SharedChannel().UpsertPost(sm.Post); err != nil {
-			scs.server.GetLogger().Log(mlog.LvlSharedChannelServiceError, "Error saving sync Post",
-				mlog.String("remote", rc.DisplayName),
-				mlog.String("ChannelId", sm.Post.ChannelId),
-				mlog.String("PostId", sm.Post.Id),
-				mlog.Err(err),
-			)
-			postErrors = append(postErrors, sm.Post.Id)
-			continue
+		// add post (may be nil if only reactions changed)
+		if sm.Post != nil {
```

Comment:
Is the order important or can this `if sm.Post != nil {` block be merged with the previous one above?

Reply:
They can be merged.  Fixed via https://github.com/mattermost/mattermost-server/pull/16783/commits/9bace6eb1613a4e5f34f48b99b964ac0fac95bab